application-admintools-default/src/main/java/com/xwiki/admintools/internal/uploadJob/UploadJobFileProcessor.java

```diff
@@ -131,40 +135,76 @@ public InputStream getArchiveInputStream(String fileRef) throws XWikiException
     }
 
     /**
-     * Process and store the content of a {@link ZipInputStream}.
+     * Process the content of a {@link ZipInputStream} and replace the old file.
      *
      * @param zis the zip content to be processed.
      * @param jobResource the resource used for storing the result.
+     * @param status the job status where to log the process result.
      * @throws IOException if any errors occur during the reading of the content.
      */
-    public void processFileContent(ZipInputStream zis, UploadPackageJobResource jobResource) throws IOException
+    public void processFileContent(ZipInputStream zis, UploadPackageJobResource jobResource,
+        PackageUploadJobStatus status) throws IOException
     {
         ByteArrayOutputStream baos = new ByteArrayOutputStream();
         byte[] buffer = new byte[1024];
         int len;
         while ((len = zis.read(buffer)) > 0) {
             baos.write(buffer, 0, len);
```

Comment:
this still loads everything to memory https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/share/classes/java/io/ByteArrayOutputStream.java#L154

you should try read your buffer, write to new location, repeat. So only 1024 bytes will be stored in memory at a time

Reply:
I also think you can have a greater size for the buffer. but i'm not entirely sure what would be a good size.